client/settings/payment-settings/account-keys-modal.js

```diff
@@ -228,45 +240,59 @@ export const AccountKeysModal = ( {
 			window.location.href = redirectOnSave;
 		} else {
 			setDisabled( false );
-			onClose();
+			onCloseHelper();
 		}
 	};
 
 	const onTabSelect = ( tabName ) => {
+		// Reset AccountKeysConnectionStatus to default state.
+		updateIsValidAccountKeys( null );
```

Comment:
One problem that prevents this from completely resetting it to its default state is that if there's still an ongoing request to validate the keys, the component can change after it has been reset 😕

We'd have to somehow cancel the request mid-flight in case it has been reset. I don't have a good suggestion on how to fix this right now, so feel free to ignore. I think it's fine as is 👍

Reply:
The same issue applies if the user closes the modal before the test is complete. 

One solution is to disable the **Cancel**, **Save test keys**, and remove the **close** button while the test is running:

![](https://cdn-std.droplr.net/files/acc_1209167/TVzVlP)

As for switching between tabs, what we can do is prevent the tab panel from switching if there's an ongoing connection test. This can be done by checking if `isTesting` validates to true. Unfortunately, we cannot prevent the tab panel tabs from thinking the tab body has in fact switched: 

![](https://cdn-std.droplr.net/files/acc_1209167/xpk7lC)

So, we may want to make this a separate issue and address it in the future. What are your thoughts on both of these solutions?

Reply:
I'm personally not a fan of blocking the UI while the test is running 🤔. I think resetting the component would be best, but it's not a big deal if that can't be accomplished in this PR.

Ideally, I think we should have the live and test keys tab work independently, and allow you to save both keys at the same time. I believe that was the original design plan, as the title used to say "Edit account keys webhooks" and the button "Save keys".

For that, we would probably have two form views rendered at the same time, and one of them hidden according to the selected tab. Similarly, the connection status component could be duplicated for test/live and that problem would be fixed.

By duplicating it, I don't mean creating a totally separate component, but rendering the same component twice with a different prop.

What do you think about that?

> So, we may want to make this a separate issue and address it in the future.

I'd also vote for addressing this in a future issue, independently of the approach chosen.

Reply:
That sounds like a good approach. That would prevent the state from leaking from **Live** into **Test**, and vice-versa.

> I'd also vote for addressing this in a future issue, independently of the approach chosen.

Sounds good. I'll go ahead and make a separate issue 😎